lib/java/src/main/java/org/apache/thrift/server/AbstractNonblockingServer.java

```diff
@@ -518,9 +522,11 @@ private boolean internalRead() {
 
     /** We're done writing, so reset our interest ops and change state accordingly. */
     private void prepareRead() {
-      // we can set our interest directly without using the queue because
-      // we're in the select thread.
-      selectionKey_.interestOps(SelectionKey.OP_READ);
+      if (selectionKey_.isValid()) {
+        // we can set our interest directly without using the queue because
+        // we're in the select thread.
+        selectionKey_.interestOps(SelectionKey.OP_READ);
+      }
       // get ready for another go-around
       buffer_ = ByteBuffer.allocate(4);
       state_ = FrameBufferState.READING_FRAME_SIZE;
```

Comment:
Not sure if the rest of `prepareRead` should go in the conditional block :thinking: